### PR TITLE
WT-6964 Refactor history store reconciliation code to use new cursor interface

### PR DESCRIPTION
**Pull request by hbokhari**

### CODE REVIEW

File: src/history/hs_rec.c [hunk 1]
```diff
@@ -964,8 +805,8 @@ __hs_fixup_out_of_order_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor,
          * case, we'll never get here.
          */
         if (insert_cursor == NULL) {
-            WT_WITHOUT_DHANDLE(session,
-              ret = __wt_open_cursor(session, WT_HS_URI, NULL, open_cursor_cfg, &insert_cursor));
+            WT_WITHOUT_DHANDLE(session, ret = __wt_curhs_open(session, NULL, &insert_cursor));
+            F_SET(insert_cursor, WT_CURSTD_HS_READ_COMMITTED);
```
Comment (quchenhao) [inline src/history/hs_rec.c:809]: For insert cursor, I don't think you need to set this flag.

File: src/history/hs_rec.c [hunk 2]
```diff
@@ -1072,19 +902,19 @@ __hs_delete_key_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, uint32_
          * The cursor will also be positioned at the start of the range that we wish to start
          * inserting.
          */
-        WT_WITHOUT_DHANDLE(session,
-          ret = __wt_open_cursor(session, WT_HS_URI, NULL, open_cursor_cfg, &insert_cursor));
+        WT_WITHOUT_DHANDLE(session, ret = __wt_curhs_open(session, NULL, &insert_cursor));
+        F_SET(insert_cursor, WT_CURSTD_HS_READ_COMMITTED);
```
Comment (quchenhao) [inline src/history/hs_rec.c:906]: Here the same

File: src/cursor/cur_hs.c [hunk 3]
```diff
@@ -381,6 +405,10 @@ __curhs_prev_visible(WT_SESSION_IMPL *session, WT_CURSOR_HS *hs_cursor)
         if (F_ISSET(std_cursor, WT_CURSTD_HS_READ_COMMITTED))
             break;
 
+        /* If the we don't have a transaction snapshot, bail out from visibility checks. */
```
Comment (quchenhao) [inline src/cursor/cur_hs.c:408]: Go back to my point in our previous discussion. If we are setting the `WT_CURSTD_HS_READ_COMMITTED` flag in many places. We can just set that flag when it is needed, in this case, for reconciliation, and remove the check for `WT_TXN_HAS_SNAPSHOT `.

**Comment by quchenhao** [comment]
@hbokhari I'm working on WT-7019, I think this needs more work when that is merged.

File: src/history/hs_rec.c (hunk 2 above)
Comment (hbokhari) [reply src/history/hs_rec.c:907]: We are calling `search_near()` on the cursor here so we have to use this flag.

Comment (quchenhao) [reply src/history/hs_rec.c:907]: I only see we do `WT_ERR(insert_cursor->insert(insert_cursor));` on the cursor. There is a search near call inside the insert implementation but that is on the underlying file cursor so should be fine.

File: src/history/hs_rec.c [hunk 4]
```diff
@@ -1072,19 +903,19 @@ __hs_delete_key_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, uint32_
...
         WT_ERR(ret);
-        F_SET(insert_cursor, WT_CURSTD_IGNORE_TOMBSTONE);
-        WT_ERR_NOTFOUND_OK(
-          __wt_hs_cursor_position(session, insert_cursor, btree_id, key, WT_TS_NONE, NULL), true);
+        insert_cursor->set_key(insert_cursor, 4, btree_id, key, WT_TS_NONE, UINT64_MAX);
+        WT_ERR_NOTFOUND_OK(insert_cursor->search_near(insert_cursor, &cmp), true);
```
Comment (hbokhari) [inline src/history/hs_rec.c:910]: @quchenhao here we are doing a `search_near()` on the `insert_cursor`.

File: src/cursor/cur_hs.c (hunk 3 above)
Comment (hbokhari) [reply src/cursor/cur_hs.c]: I have replaced this condition with an assert - if someone using the HS cursor doesn't have the snapshot, they should set the appropriate flag.

File: src/cursor/cur_hs.c [hunk 5]
```diff
@@ -465,6 +509,11 @@ __curhs_next_visible(WT_SESSION_IMPL *session, WT_CURSOR_HS *hs_cursor)
         if (F_ISSET(std_cursor, WT_CURSTD_HS_READ_COMMITTED))
             break;
 
+        /*
```
Comment (quchenhao) [inline src/cursor/cur_hs.c:512]: Can it be a one line comment?

File: src/cursor/cur_hs.c [hunk 6]
```diff
@@ -481,6 +530,49 @@ __curhs_next_visible(WT_SESSION_IMPL *session, WT_CURSOR_HS *hs_cursor)
...
+    WT_CURSOR_HS *hs_cursor;
+    WT_DECL_ITEM(srch_key);
+    WT_DECL_RET;
+    int cmp;
+
+    hs_cursor = (WT_CURSOR_HS *)cursor;
+    WT_ERR(__wt_scr_alloc(session, 0, &srch_key));
+    WT_ERR(__wt_buf_set(session, srch_key, cursor->key.data, cursor->key.size));
+    WT_ERR(cursor->search_near(cursor, &cmp));
+    if (F_ISSET(hs_cursor, WT_HS_CUR_COUNTER_SET)) {
```
Comment (quchenhao) [inline src/cursor/cur_hs.c:549]: I think we should break up this function into two separate helper functions. One to place the cursor before the search key and the other after the search key. We should not decide where to place the key based on the WT_HS_CUR_COUNTER_SET flag.

File: src/cursor/cur_hs.c [hunk 7]
```diff
@@ -511,133 +606,99 @@ __curhs_search_near(WT_CURSOR *cursor, int *exactp)
     WT_ERR(__wt_hs_cursor_search_near(session, file_cursor, &exact));
 
     /*
-     * There are some key fields missing so we are searching a range of keys. Place the cursor at
-     * the start of the range.
+     * If we have some missing fields in the key, try to walk forward first to increase the chance
```
Comment (quchenhao) [inline src/cursor/cur_hs.c:609]: The comment is obsolete. I forgot to remove it. We should rephrase or remove it.

File: src/history/hs_rec.c [hunk 8]
```diff
@@ -878,15 +702,19 @@ __hs_fixup_out_of_order_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor,
      * to keep doing "next" until we've got a key greater than the one we attempted to position
      * ourselves with.
      */
-    for (; ret == 0; ret = __wt_hs_cursor_next(session, hs_cursor)) {
+    for (; ret == 0; ret = hs_cursor->next(hs_cursor)) {
         /*
          * Prior to getting here, we've done a "search near" on our key for the timestamp we're
```
Comment (quchenhao) [inline src/history/hs_rec.c:707]: These comments are obsolete.

File: src/history/hs_rec.c [hunk 9]
```diff
@@ -878,15 +702,19 @@ __hs_fixup_out_of_order_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor,
...
          * inserting and then a "next". In the regular case, our cursor will be positioned on the
          * next key and we'll break out of the first iteration in one of the conditions below.
          */
         WT_ERR(hs_cursor->get_key(hs_cursor, &hs_btree_id, &hs_key, &hs_ts, &hs_counter));
-        WT_ERR(__wt_compare(session, NULL, &hs_cursor->key, srch_key, &cmp));
-        if (cmp > 0)
+        if (hs_btree_id != btree->id)
```
Comment (quchenhao) [inline src/history/hs_rec.c:712]: I don't think we need to check this as we should get not found if btree id isn't a match. We can turn this into an assert.

File: src/history/hs_rec.c [hunk 10]
```diff
@@ -878,15 +702,19 @@ __hs_fixup_out_of_order_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor,
...
+            break;
+        if (hs_ts > ts)
+            break;
+        WT_ERR(__wt_compare(session, NULL, &hs_key, key, &cmp));
```
Comment (quchenhao) [inline src/history/hs_rec.c:716]: Same for key. We can make this an assert instead.

File: src/history/hs_rec.c [hunk 11]
```diff
@@ -913,7 +741,7 @@ __hs_fixup_out_of_order_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor,
      * 2     foo 3  2       ccc
      * 2     foo 3  3       ddd
      */
-    for (; ret == 0; ret = __wt_hs_cursor_next(session, hs_cursor)) {
+    for (; ret == 0; ret = hs_cursor->next(hs_cursor)) {
         /*
          * Prior to getting here, we've done a "search near" on our key for the timestamp we're
```
Comment (quchenhao) [inline src/history/hs_rec.c:746]: Also obsolete

File: src/history/hs_rec.c [hunk 12]
```diff
@@ -926,15 +754,7 @@ __hs_fixup_out_of_order_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor,
         WT_ERR(__wt_compare(session, NULL, &hs_key, key, &cmp));
         if (cmp != 0)
```
Comment (quchenhao) [inline src/history/hs_rec.c:755]: We should make the btree id check and the key check as asserts.

File: src/cursor/cur_hs.c (hunk 5 above)
Comment (hbokhari) [reply src/cursor/cur_hs.c:512]: Converting this to a single line comment overflows it.

File: src/cursor/cur_hs.c (hunk 6 above)
Comment (hbokhari) [reply src/cursor/cur_hs.c:549]: Not sure on this one - we have been deliberately overloading the HS cursor interface calls based on what fields are set. Breaking up the helper function will probably make it more confusing.

Comment (quchenhao) [reply src/cursor/cur_hs.c:549]: The helper function I think itself is confusing. It is a bit dangerous to assume that if the counter is set, then we want our cursor to be placed on the smaller key. It is the same problem as implementing this in search near itself. It should be the caller who decides where they want to place the cursor at.

File: src/history/hs_rec.c (hunk 8 above)
Comment (hbokhari) [reply src/history/hs_rec.c:707]: Fixed.

File: src/history/hs_rec.c (hunk 9 above)
Comment (hbokhari) [reply src/history/hs_rec.c:712]: Fixed.

File: src/history/hs_rec.c (hunk 10 above)
Comment (hbokhari) [reply src/history/hs_rec.c:716]: Fixed.

File: src/history/hs_rec.c (hunk 11 above)
Comment (hbokhari) [reply src/history/hs_rec.c:746]: Fixed.

File: src/cursor/cur_hs.c (hunk 6 above)
Comment (hbokhari) [reply src/cursor/cur_hs.c]: Makes sense. I have divided the helper function into distinct and more clearly named functions.

**Review by quchenhao — APPROVED** [review]
lgtm

**Review by bvpvamsikrishna — APPROVED** [review]
lgtm

**Comment by hbokhari** [comment]
retest this please

**Comment by hbokhari** [comment]
evergreen retry

**Pull request merged**